Trees/BinarySearchTree.cpp

```diff
@@ -0,0 +1,348 @@
+/*
+    Author:
+    Varun Jayathirtha (www.github.com/varunjm)
+
+    Binary Search Tree:
+    A graph data structure where each node has a unidirectional path to at most
+    two node, none of which are itself or its predecessors, i.e the graph is
+    acyclic. The first node is called the root node. Another important feature
+    of this tree is that each node has values lesser than or equal to itself in
+    the left subtree and values greater than or equal to itself in the right
+    subtree.
+*/
+
+#include "BinarySearchTree.hpp"
+#include <iostream>
+#include <climits>
+#include <stack>
+
+using namespace std;
+
+Node * get_node() {
+    Node * newNode = new Node();
+    newNode->left_child = newNode->right_child = nullptr;
+
+    return newNode;
+}
+
+BinarySearchTree :: BinarySearchTree() {
+    root = get_node();
+    root->value = INT_MAX;
+}
+
+/*
+    Insert
+    This traverses the tree to find the right spot to insert the newly given
+    value. At each node, it checks if the node has a value lesser than or equal
+    to itself, if so it goes to its left subtree. Otherwise it goes to the right
+    subtree. When it meets an end point i.e. NULL it inserts the new Node along
+    with the new value
+
+    Time complexity:
+    Best case : O(log n), where n is the number of nodes in the tree
+    Worst case : O(n)
+
+    Space complexity:
+    O(1)
+*/
+
+bool BinarySearchTree :: insert_node(int value) {
+    // If first insertion, replace the value of the already created node
+    if(root->value == INT_MAX) {
+        root->value = value;
+        return true;
+    } else {
+        Node * iter = root, * newNode = get_node();
```

Comment:
Avoid declaring multiple pointers in the same declaration. [See details](https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions).